### ISSUE TITLE
concatenating `tf_reg`s to produce a `tf_irreg`

### ISSUE THREAD

**Issue by jeff-goldsmith** [tidyfun/tidyfun#103]
Concatenating two `tf_reg`s that are observed over different domains should (and until `vctrs`, did ...) produce a new `tf_irreg` vector, with the `arg`s for each `tf_reg` retained. That's broken now:

``` r
library(tidyfun)
#> Registered S3 method overwritten by 'GGally':
#>   method from   
#>   +.gg   ggplot2
#> 
#> Attaching package: 'tidyfun'
#> The following objects are masked from 'package:stats':
#> 
#>     sd, var

x = tf_rgp(1, arg = seq(0, 1, l = 11))
y = tf_rgp(1, arg = seq(0, 1, l = 21))

c(x, y)
#> Warning in vec_ptype2.tfd.tfd(x = x, y = y, x_arg = x_arg, y_arg = y_arg):
#> concatenating functions on different grids.
#> Warning in vec_ptype2.tfd.tfd(x = x, y = y, x_arg = x_arg, y_arg = y_arg):
#> inputs have different resolutions, result has resolution =0.01
#> tfd[2] on (0,1)
#> Error in .x$value: $ operator is invalid for atomic vectors
```

<sup>Created on 2020-04-22 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

I *think* this is because `vctrs` is using `vec_data` to strip out attributes (before `vec_ptype2.tfb.tfb`?), and then `vec_restore` is trying to put the attributes back in. But `tf_reg` and `tf_irreg` have different attribute structures, because `tf_reg` uses the same `arg`s for all functions, so we're slapping some `tf_reg` attributes on a `tf_irreg` output. 

It's maybe possible to have a smarter `vec_restore`, but I haven't figured that out yet. 

It is possible to make `tf_reg` more like `tf_irreg` -- but then we lose all the memory efficiency that was the motivation for having these separate in the first place, and I think @fabian-s would go slightly nuts.

**Comment by jeff-goldsmith** [comment]
An additional thought -- since `vctrs` implements some pretty specific rules, I think we're going to have a real challenge keeping `tfd_reg` and `tfd_irreg`. 

Right now, we sort of want `c(x, y)` to return outputs with different classes when `x` and `y` are both `tfd_reg`, depending on whether they have the same `arg` attribute -- if they do we want a `tfd_reg`, and if the don't we want a `tfd_irreg`. I still don't fully understand `vctrs`, but I think that's something they want to avoid. 

Maybe it's possible to always return a `tfd_irreg`, and then simplify that to `tfd_reg` when the `arg`s are identical?

**Comment by jeff-goldsmith** [comment]
closed by 8548c40cc4f972cd0494c7d40f8debe28b48b88b

mostly needed a bunch more `vec_cast`s and `vec_ptype2`s, along with some handling of different cases.

also note that returning different output (sub)classes is sort of against the `vctrs` design philosophy, but oh well
